drivers/nhs-nut.c

```diff
@@ -0,0 +1,2070 @@
+                strcpy(data.upsdesc,"LASER ONLINE 5000VA");
+                data.VA = 5000;
+        break;
+
+        case 113:
+                data.upscode = 113;
+                strcpy(data.upsdesc,"PDV SENOIDAL ISOLADOR 700VA");
+                data.VA = 700;
+        break;
+
+        default:
+                data.upscode = -1;
+                strcpy(data.upsdesc,"NHS UNKNOWN");
+                data.VA = 0;
+        break;
+    }
+    return data;
+}
+
+unsigned int get_va(int equipment) {
+    upsinfo ups;
+    char * va = getval("va");
+    ups = getupsinfo(equipment);
+    if (ups.VA > 0)
+        return ups.VA;
+    else {
+        if (va)
+            return atoi(va);
+        else
+            fatalx(EXIT_FAILURE,"Please set VA (Volt Ampere) nominal capacity value to your equipment in ups.conf.");
+    }
+}
+
+float get_pf() {
+    char * pf = getval("pf");
+    if (pf)
+        return atof(pf);
+    else
+        return DEFAULTPF;
+}
+
+unsigned int get_ah() {
+    char * ah = getval("ah");
+    if (ah)
+        return (unsigned int)atoi(ah);
+    else
+        fatalx(EXIT_FAILURE,"Please set AH (Ampere Hour) value to your battery's equipment in ups.conf.");
+}
+
+float get_vin_perc(char * var) {
+    char * perc = getval(var);
+    if (perc)
+        return atof(perc);
+    else
+        return DEFAULLTPERC;
+}
+
+void upsdrv_initinfo(void) {
+    char * b = getval("baud");
+    upsdebugx(1,"Port is %s and baud_rate is %s",device_path,b);
+    baudrate = DEFAULTBAUD;
+    upsdebugx(1,"Start initinfo()");
+    if (b)
+        baudrate = atoi(b);
+    if (device_path) {
+        if (strcasecmp(device_path,"auto") == 0)
+            strcpy(porta,DEFAULTPORT);
+        else
+            strcpy(porta,device_path);
+        serial_fd = openfd(porta,baudrate);
+        if (serial_fd == -1)
+            fatalx(EXIT_FAILURE,"Unable to open port %s with baud %d",porta,baudrate);
+        else {
+            upsdebugx(1,"Communication started on port %s, baud rate %d. Calling updateinfo()",porta,baudrate);
+        }
+    }
+    else
+        fatalx(EXIT_FAILURE,"Unable to define port and baud");
+    upsdebugx(1,"End initinfo()");
+}
+
+float calculate_efficiency(float vacoutrms, float vacinrms) {
+    return (vacoutrms * vacinrms) / 100.0;
+}
+
+void upsdrv_updateinfo(void) {
+    // retries to open port
+    unsigned int retries = 3;
+    unsigned int i = 0;
+    char alarm[1024];
+    unsigned int va = 0;
+    unsigned int ah = 0;
+    unsigned int vin_underv = 0;
+    unsigned int vin_overv = 0;
+    unsigned int perc = 0;
+    unsigned int vin = 0;
+    unsigned int vout = 0;
+    unsigned int autonomy_secs = 0;
+    float vin_low_warn = 0;
+    float vin_low_crit = 0;
+    float vin_high_warn = 0;
+    float vin_high_crit = 0;
+    float calculated = 0;
+    float vpower = 0;
+    float pf = 0;
+    int bcharge = 0;
+    int min_input_power = 0;
+    
+    upsinfo ups;
+
+    upsdebugx(1,"Start updateinfo()");
```

Comment:
Generally (in many methods), for methodical tracing you can use `__func__` macro, e.g. `upsdebugx(4, "%s: starting...", __func__);` is a common sight in NUT code (maybe with parentheses - more often to include significant argument values in the log).